PWGCF/MultiparticleCorrelations/Core/MuPa-MemberFunctions.h

```diff
@@ -2015,4 +2253,210 @@ Double_t Weight(const Double_t& value,
+    reachedMaxNumberOfEvents = kTRUE;
+  }
+
+  return reachedMaxNumberOfEvents;
+
+} // void MaxNumberOfEvents()
+
+//============================================================
+
+template <typename T>
+void MainLoopOverParticles(T const& tracks, eRecSim rs)
+{
+  // This is the main loop over particles, in which Q-vectors and particle histograms are filled, particle cuts applied, etc.
+  // If I want to access both rec and sim info in this loop, set gProcessRecSim = true via configurable "cfWhatToProcess"
+
+  if (fVerbose) {
+    LOGF(info, "\033[1;32m%s\033[0m", __PRETTY_FUNCTION__);
+  }
+
+  Double_t dPhi = 0., wPhi = 1.; // azimuthal angle and corresponding phi weight
+  Double_t dPt = 0., wPt = 1.;   // transverse momentum and corresponding pT weight
+  Double_t dEta = 0., wEta = 1.; // pseudorapidity and corresponding eta weight
+  Double_t wToPowerP = 1.;       // weight raised to power p
+  fSelectedTracks = 0;           // reset number of selected tracks
+  for (auto& track : tracks) {
+
+    // *) Fill particle histograms for reconstructed particle before particle cuts:
+    FillParticleHistograms(track, rs, eBefore);
+
+    // *) Fill particle histograms for corresponding simulated particle before particle cuts:
+    if (gProcessRecSim) {
+      if (track.has_mcParticle()) {
+        auto mcParticle = track.mcParticle();
+        FillParticleHistograms(mcParticle, eSim, eBefore); // great, this works straight like this, thanks O2 people!
+      } else {
+        LOGF(debug, "in function \033[1;31m%s at line %d, track.has_mcParticle() is false. \033[0m", __PRETTY_FUNCTION__, __LINE__);
+      }
+    } // if (gProcessRecSim) {
+
+    // *) Particle cuts:
+    if (!ParticleCuts(track, rs)) {
+      continue;
+    }
+
+    // *) Fill particle histograms for reconstructed data after particle cuts:
+    FillParticleHistograms(track, rs, eAfter);
+
+    // *) Fill particle histograms for corresponding simulated particle after particle cuts:
+    // TBI 20231020 there is a duplication of code here, see above. I should get 'mcParticle' only once for this 'track' in this loop
+    if (gProcessRecSim) {
+      if (track.has_mcParticle()) {
+        auto mcParticle = track.mcParticle();
+        FillParticleHistograms(mcParticle, eSim, eAfter); // great, this works straight like this, thanks O2 people!
+      } else {
+        LOGF(debug, "in function \033[1;31m%s at line %d, track.has_mcParticle() is false. \033[0m", __PRETTY_FUNCTION__, __LINE__);
+      }
+    } // if (gProcessRecSim) {
+
+    // *) Fill Q-vectors:
+    if (gProcessRec || gProcessRecSim) {
+      //  Take kinematics from reconstructed particles:
+      dPhi = track.phi();
+      dPt = track.pt();
+      dEta = track.eta();
+    } else if (gProcessSim) {
+      //  Take kinematics from simulated particles:
+      dPhi = track.mcParticle().phi();
+      dPt = track.mcParticle().pt();
+      dEta = track.mcParticle().eta();
```

Comment:
This will not compile either. See above comment

Reply:
Yes, see the above reply.